File: spaniel.py
#!/usr/bin/python

"""Query command implementation for mutt, searching a given vcard directory

Usage:
    spaniel.py [--abook=<abook>] [<query_string>]


Options:
    --abook=<abook>  The name of the VCF file or directory to search [default: .].
    <query_string>   The string to search for.
"""

from os import listdir
from os.path import isdir, isfile, join
import sys

from docopt import docopt
import vobject


def vcard_match(vcard, query):
    return query.upper() in str(vcard).upper()


def entry_output(email, name):
    sys.stdout.write(email.strip())
    sys.stdout.write('\t')
    sys.stdout.write(name.strip())
    sys.stdout.write('\t\n')


def process_vcard(vcard, query):
    if vcard_match(vcard, query):
        name = '<no name>'
        if hasattr(vcard, 'n'):
            name = str(vcard.n.value)

        if hasattr(vcard, 'email_list'):
            for email in vcard.email_list:
                entry_output(email.value, name)
        elif hasattr(vcard, 'email'):
            entry_output(vcard.email.value, name)


def query_directory(dirname, query):

    for f in listdir(dirname):
        filename = join(dirname, f)

        if isfile(filename):
            try:
                vcard = vobject.readOne(open(filename, 'r'))
                process_vcard(vcard, query)
            except vobject.base.ParseError:
                pass


def query_file(filename, query):
    for vcard in vobject.base.readComponents(open(filename, 'r')):
        process_vcard(vcard, query)


if __name__ == '__main__':
    arguments = docopt(__doc__, version='v0.1')

    abook = arguments['--abook']
    query = arguments['<query_string>']

    # First line of output ignored
    print('Searching...')

    if query is not None and len(query) > 0:
        if isfile(abook):
            query_file(abook, query)
        elif isdir(abook):
            query_directory(abook, query)

File: spaniel.js
#!/usr/bin/node

const fs = require('fs')
const path = require('path')
const util = require('util')

const commandLineArgs = require('command-line-args')
const vdata = require('vdata-parser')

const EMAIL = 'EMAIL'
const FN = 'FN'
const PHOTO = 'PHOTO'
const TYPE = 'TYPE'
const VCARD = 'VCARD'


function processCommandLineOrFail() {
    const optionDefinitions = [
      { name: 'abook',
        alias: 'a',
        type: String,
        defaultOption: false,
        defaultValue: '.'},
      { name: 'query',
        alias: 'q',
        type: String,
        defaultOption: true },
      { name: 'help',
        alias: 'h',
        type: Boolean }
    ]

    const options = commandLineArgs(optionDefinitions)

    if (!options.query || options.help) {
        console.log('Usage: spaniel.js [--abook <file/dir>] <query>')
        console.log('')
        console.log('    --abook <file/dir> the file or directory containing vCards')
        console.log('    <query>            a string to search for in the vCards')
        process.exit(options.help ? 0 : -1)
    }

    return options
}


function isMatch(vc, query) {
    const replacer = (key, value) => {
        if (key == PHOTO)
            return ''
        else
            return value
    }
    const vcString = JSON.stringify(vc, replacer)
    return vcString.search(new RegExp(query, 'i')) != -1
}


// fullname is a string
// email is a email object from vdata-parser
function outputEmailHit(fullname, email) {
    var type = ''
    var emailString = email

    if (email.params) {
        email.params.forEach(param => {
            if (TYPE in param) {
                type = param[TYPE]
            }
        })
        emailString = email.value
    }

    console.log(emailString + '\t' + fullname + '\t' + type)
}


function outputVcardHit(vc) {
    if (EMAIL in vc) {
        const emails = vc[EMAIL]
        if (emails.forEach) {
            emails.forEach(email => {
                outputEmailHit(vc[FN], email)
            })
        } else {
            outputEmailHit(vc[FN], emails)
        }
    }
}


function processVcard(vc, query) {
    if (isMatch(vc, query) &&true) {
        outputVcardHit(vc)
    }
}


async function searchFile(abook, query) {
    vdata.promFromFile = util.promisify(vdata.fromFile)
    const data = await vdata.promFromFile(abook)

    if (data[VCARD]) {
        if (data[VCARD].forEach) {
            data[VCARD].forEach(vc => {
                processVcard(vc, query)
            })
        } else {
            processVcard(data[VCARD], query)
        }
    } else {
        throw "Card is not a vCard"
    }
}


async function searchDirectory(abook, query) {
    const readdir = util.promisify(fs.readdir)

    const files = await readdir(abook)
    files.forEach(file => {
        const fullname = path.join(abook, file)
        searchFile(fullname, query).catch(err => {
            // Ignore error messages
        })
    })
}


async function doSearch(abook, query) {
    const lstat = util.promisify(fs.lstat)

    const stats = await lstat(abook)

    if (stats.isFile()) {
        await searchFile(abook, query)
    } else if (stats.isDirectory()) {
        await searchDirectory(abook, query)
    } else {
        throw abook + ' is not a file or directory'
    }
}


const options = processCommandLineOrFail()
doSearch(options.abook, options.query).catch(err => {
    console.error(err.message)
    process.exit(-1)
})



